File: scripts/import-batch-house-staging.py
#!/usr/bin/env python3
"""Import Batch House equipment TSV into staging_objects SQL migration."""
import json
import re
from datetime import date
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
INPUT = ROOT / "data" / "batch-house-equipment.tsv"
OUTPUT = ROOT / "supabase" / "migrations" / "0012_batch_house_staging.sql"

# Full legend: prefix -> type_code  (None = awaiting clarification from user)
LEGEND = {
    "BA":   "bin_activator",         # Bin Activator
    "BE":   "bucket_elevator",       # Bucket Elevator
    "C":    "conveyor",              # Conveyor Belt
    "DCC":  "pneumatic_hammer",      # Pneumatic Hammer
    "DCU":  "dedusting_unit",        # Dedusting Unit
    "DE":   None,                    # Unknown — needs clarification
    "DG":   "diverter_gate",         # Diverter Gate
    "FFS":  None,                    # Unknown — needs clarification
    "HM":   "hose_mover",            # Hose Mover
    "HO":   "receiving_hopper",      # Receiving Hopper
    "M":    "mixer",                 # Mixer
    "MI":   "premixer",              # Premixer
    "MMU":  "moisture_probe",        # Moisture Probe
    "MS":   "moisture_shutoff",      # Moisture Shut-Off Valve
    "MW":   None,                    # Unknown — needs clarification
    "PM":   None,                    # Unknown — needs clarification
    "PTP":  "pneumatic_pipeline",    # Pneumatic Transfer Pipeline (sender)
    "S":    "silo_accessory",        # Silo Accessory
    "SC":   "dosing_screw",          # Dosing Screw
    "TB":   "terminal_block",        # Terminal Block
    "TW":   "test_weight",           # Automatic Test Weight
    "VF":   "dosing_feeder",         # Dosing Vibratory Feeder
}

# Order matters — longer prefixes first to avoid mis-matching e.g. "DCC" as "DC"
_PREFIX_ORDER = [
    ("DCC", "DCC"),
    ("DCU", "DCU"),
    ("DG",  "DG"),
    ("DE",  "DE"),
    ("FFS", "FFS"),
    ("HM",  "HM"),
    ("HO",  "HO"),
    ("MMU", "MMU"),
    ("MS",  "MS"),
    ("MW",  "MW"),
    ("MI",  "MI"),
    ("PTP", "PTP"),
    ("PM",  "PM"),
    ("SC",  "SC"),
    ("TB",  "TB"),
    ("TW",  "TW"),
    ("VF",  "VF"),
    ("BE",  "BE"),
    ("BA",  "BA"),
    ("CG",  None),   # CG = valve/gate families, keep as process_equipment
    ("CC",  None),
    ("CT",  None),
    ("C",   "C"),    # plain conveyor — only if starts with C + digit
    ("M",   "M"),
    ("S",   "S"),
]


def infer_type_code(tag: str) -> str:
    # Use the full tag (before first '-') as the base
    base = tag.split("-")[0].upper()
    for prefix, legend_key in _PREFIX_ORDER:
        if base.startswith(prefix) and legend_key is not None:
            # Special case: "C" prefix only matches C<digit> conveyors
            if prefix == "C" and not re.match(r"^C[0-9]", base):
                continue
            # Special case: "M" prefix — bare Mx or Mx#### pattern only
            if prefix == "M" and not re.match(r"^M[0-9]", base):
                continue
            # Special case: "S" prefix — silo accessories S<digit> or S3…
            if prefix == "S" and not re.match(r"^S[0-9]", base):
                continue
            tc = LEGEND.get(legend_key)
            return tc if tc else "process_equipment"
        elif base.startswith(prefix) and legend_key is None:
            return "process_equipment"
    return "process_equipment"


def elevation_to_floor(elev: float, subtitle: str) -> str:
    if elev <= 0:
        return "Ground"
    if elev <= 11:
        return "Level 5"
    if elev <= 14:
        return "Level 12"
    if elev <= 20:
        return "Level 17"
    if elev <= 25:
        return "Level 22"
    if elev <= 30:
        return "Level 27"
    return "Level 32"


def parse_level(level_raw: str, subtitle: str):
    if not level_raw or level_raw.strip() in ("", "5"):
        return None, "Ground", None
    s = level_raw.strip()
    if s == "CULLET RETURN":
        return 0.0, "Ground", "cullet_return"

    m = re.match(r"([+-]?\d+(?:\.\d+)?)\.\.\.([+-]?\d+(?:\.\d+)?)", s)
    if m:
        mid = (float(m.group(1)) + float(m.group(2))) / 2
        return mid, elevation_to_floor(mid, subtitle), infer_zone(subtitle)

    dual = re.search(r"([+-]?\d+(?:\.\d+)?)\s*/\s*([+-]?\d+(?:\.\d+)?)", subtitle)
    try:
        n = float(s.replace("+", ""))
    except ValueError:
        return None, "Ground", infer_zone(subtitle)

    if dual and (n < 5 or n > 40):
        mid = (float(dual.group(1)) + float(dual.group(2))) / 2
        return mid, elevation_to_floor(mid, subtitle), infer_zone(subtitle)

    return n, elevation_to_floor(n, subtitle), infer_zone(subtitle)


def infer_zone(subtitle: str) -> str:
    if "CULLET TOWER" in subtitle:
        return "cullet_tower"
    if "BATCH TRANSPORT START" in subtitle:
        return "batch_transport_start"
    if "BATCH TRANSPORT END" in subtitle:
        return "batch_transport_end"
    if "CULLET RETURN" in subtitle:
        return "cullet_return"
    return "batch_house"


_TYPE_LABELS = {
    "bin_activator":    "Bin Activator",
    "bucket_elevator":  "Bucket Elevator",
    "conveyor":         "Conveyor",
    "pneumatic_hammer": "Pneumatic Hammer",
    "dedusting_unit":   "Dedusting Unit",
    "diverter_gate":    "Diverter Gate",
    "hose_mover":       "Hose Mover",
    "receiving_hopper": "Receiving Hopper",
    "mixer":            "Mixer",
    "premixer":         "Premixer",
    "moisture_probe":   "Moisture Probe",
    "moisture_shutoff": "Moisture Shut-Off Valve",
    "pneumatic_pipeline": "Pneumatic Transfer Pipeline",
    "silo_accessory":   "Silo Accessory",
    "dosing_screw":     "Dosing Screw",
    "terminal_block":   "Terminal Block",
    "test_weight":      "Automatic Test Weight",
    "dosing_feeder":    "Dosing Vibratory Feeder",
    "process_equipment": "Process Equipment",
}


def build_name(tag: str, subtitle: str, type_code: str) -> str:
    label = _TYPE_LABELS.get(type_code, "Equipment")
    zone = subtitle.split(" +")[0].strip()
    return f"{label} {tag} — {zone}"


def esc(s) -> str:
    return str(s or "").replace("'", "''")


def parse_tsv(text: str):
    rows = []
    for line in text.splitlines():
        if not line.strip():
            continue
        cols = line.split("\t")
        if len(cols) < 2:
            continue
        tag = cols[0].strip()
        if tag in ("tag", "5", ""):
            continue
        rows.append(
            {
                "tag": tag,
                "area_code": cols[1].strip() if len(cols) > 1 else "",
                "subtitle_location": cols[2].strip() if len(cols) > 2 else "",
                "drawing": cols[3].strip() if len(cols) > 3 else "",
                "level": cols[4].strip() if len(cols) > 4 else "",
                "project_site": cols[5].strip() if len(cols) > 5 else "",
            }
        )
    return rows


def dedupe_tags(rows):
    seen = {}
    out = []
    for row in rows:
        key = row["tag"]
        unique = key
        note = None
        if key in seen:
            prev = seen[key]
            suffix = row["drawing"] or row["level"] or str(len(seen))
            unique = f"{key}@{suffix}"
            note = f"Duplicate tag; also at drawing {prev['drawing']} level {prev['level']}"
        row["unique_tag"] = unique
        row["source_notes"] = note
        seen[key] = row
        out.append(row)
    return out


def main():
    rows = dedupe_tags(parse_tsv(INPUT.read_text(encoding="utf-8")))
    print(f"Parsed {len(rows)} rows")

    values = []
    for r in rows:
        type_code = infer_type_code(r["tag"])
        elev, floor, zone = parse_level(r["level"], r["subtitle_location"])
        zone = zone or infer_zone(r["subtitle_location"])
        props = {
            "area_code": r["area_code"],
            "subtitle_location": r["subtitle_location"],
            "drawing": r["drawing"],
            "level_raw": r["level"],
            "elevation_m": elev,
            "zone": zone,
            "project_site": r["project_site"],
            "original_tag": r["tag"],
            "location_code": f"BAH-{elev if elev is not None else 0}-{zone.upper()}",
        }
        if type_code == "process_equipment":
            props["legend_pending"] = True
        props_json = esc(json.dumps(props, ensure_ascii=False))
        note_sql = f"'{esc(r['source_notes'])}'" if r["source_notes"] else "NULL"
        values.append(
            f"  ('{esc(r['unique_tag'])}','{esc(r['tag'])}','{esc(build_name(r['tag'], r['subtitle_location'], type_code))}',"
            f"'{type_code}','BTH-03','{floor}','{props_json}'::jsonb,"
            f"'batch-house-equipment-export',{note_sql},'medium','draft')"
        )

    sql = f"""-- Migration 0012: Batch House process equipment → staging_objects
-- Source: batch-house-equipment.tsv
-- Legend: BA=Bin Activator, BE=Bucket Elevator, C=Conveyor, DCC=Pneumatic Hammer
-- Generated: {date.today().isoformat()}

INSERT INTO public.staging_objects
  (tag, proposed_code, name, type_code, building_code, floor_name, properties, source_doc, source_notes, confidence, status)
VALUES
{',\n'.join(values)}
ON CONFLICT (tag) DO UPDATE SET
  name = EXCLUDED.name,
  type_code = EXCLUDED.type_code,
  building_code = EXCLUDED.building_code,
  floor_name = EXCLUDED.floor_name,
  properties = EXCLUDED.properties,
  source_doc = EXCLUDED.source_doc,
  source_notes = COALESCE(EXCLUDED.source_notes, staging_objects.source_notes),
  updated_at = now();
"""
    OUTPUT.write_text(sql, encoding="utf-8")
    print(f"Wrote {OUTPUT} ({len(values)} rows, {OUTPUT.stat().st_size // 1024} KB)")


if __name__ == "__main__":
    main()

File: scripts/import-batch-house-staging.mjs
/**
 * Import Batch House process equipment tags into staging_objects.
 * Usage: node scripts/import-batch-house-staging.mjs [path-to.tsv]
 * Default input: data/batch-house-equipment.tsv
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.join(__dirname, '..');
const inputPath = process.argv[2] || path.join(root, 'data', 'batch-house-equipment.tsv');
const outSql = path.join(root, 'supabase', 'migrations', '0012_batch_house_staging.sql');

const LEGEND = {
  BA: 'bin_activator',
  BE: 'bucket_elevator',
  C: 'conveyor',
  DCC: 'pneumatic_hammer',
};

function inferTypeCode(tag) {
  const base = tag.split('-')[0].toUpperCase();
  if (base.startsWith('DCC')) return LEGEND.DCC;
  if (base.startsWith('BE')) return LEGEND.BE;
  if (base.startsWith('BA')) return LEGEND.BA;
  // Conveyor tags: C1121, C6131A, C9281A (not CG, CC, CT equipment families)
  if (/^C[0-9]/.test(base) && !/^(CG|CC|CT|CP|C9)/.test(base)) return LEGEND.C;
  return 'process_equipment';
}

function inferEquipmentClass(tag) {
  const suffix = tag.includes('-') ? tag.split('-').slice(1).join('-').toUpperCase() : '';
  if (/^XV|^ZS|^JB|^CB|^CU|^EU|^FD|^MP|^FM|^WE|^DP|^LP|^M\d|^Q\d|^S51|^SS|^AS|^ES|^MS|^HS|^TC|^LS|^OT|^G\d|^B\d|^H\d|^D\d|^U\d|^P\d|^SST|^MGC|^MGE|^M2X|^CUT|^AHL|^AHS|^AHG|^LCP|^SLS|^LSL|^CV|^DPSS|^PS|^SM|^FM|^JB|^WE/.test(suffix)) {
    return suffix.replace(/\d+$/, '').slice(0, 6) || 'component';
  }
  return null;
}

function parseLevel(levelRaw, subtitle) {
  if (!levelRaw || levelRaw === '5' || levelRaw.trim() === '') return { elevation_m: null, floor: 'Ground' };
  const s = String(levelRaw).trim();
  if (s === 'CULLET RETURN') return { elevation_m: 0, floor: 'Ground', zone: 'cullet_return' };

  // Range like +8.190...+9.965 or +32.460...+42.125
  const range = s.match(/([+-]?\d+(?:\.\d+)?)\.\.\.([+-]?\d+(?:\.\d+)?)/);
  if (range) {
    const mid = (parseFloat(range[1]) + parseFloat(range[2])) / 2;
    return { elevation_m: mid, floor: elevationToFloor(mid, subtitle) };
  }

  // Dual elevation in subtitle: +26.300 / +29.000 — use subtitle if level looks like ratio
  const dual = subtitle.match(/([+-]?\d+(?:\.\d+)?)\s*\/\s*([+-]?\d+(?:\.\d+)?)/);
  if (dual && (parseFloat(s) < 5 || parseFloat(s) > 40)) {
    const mid = (parseFloat(dual[1]) + parseFloat(dual[2])) / 2;
    return { elevation_m: mid, floor: elevationToFloor(mid, subtitle) };
  }

  const n = parseFloat(s.replace(/^\+/, ''));
  if (Number.isNaN(n)) return { elevation_m: null, floor: 'Ground' };
  return { elevation_m: n, floor: elevationToFloor(n, subtitle) };
}

function elevationToFloor(elev, subtitle) {
  if (subtitle.includes('CULLET TOWER')) {
    if (elev <= 0) return 'Ground';
    if (elev <= 6) return 'Level 5';
    if (elev <= 14) return 'Level 12';
    if (elev <= 20) return 'Level 17';
    if (elev <= 25) return 'Level 22';
    if (elev <= 30) return 'Level 27';
    return 'Level 32';
  }
  if (elev <= 0) return 'Ground';
  if (elev <= 6) return 'Level 5';
  if (elev <= 11) return 'Level 5'; // +10.435 areas
  if (elev <= 14) return 'Level 12';
  if (elev <= 20) return 'Level 17';
  if (elev <= 25) return 'Level 22';
  if (elev <= 30) return 'Level 27';
  return 'Level 32';
}

function inferZone(subtitle) {
  if (subtitle.includes('CULLET TOWER')) return 'cullet_tower';
  if (subtitle.includes('BATCH TRANSPORT START')) return 'batch_transport_start';
  if (subtitle.includes('BATCH TRANSPORT END')) return 'batch_transport_end';
  if (subtitle.includes('CULLET RETURN')) return 'cullet_return';
  if (subtitle.includes('BATCH HOUSE')) return 'batch_house';
  return 'batch_house';
}

function inferBuilding(subtitle) {
  return 'BTH-03';
}

function buildName(tag, subtitle, typeCode) {
  const typeLabel = {
    bin_activator: 'Bin Activator',
    bucket_elevator: 'Bucket Elevator',
    conveyor: 'Conveyor',
    pneumatic_hammer: 'Pneumatic Hammer',
    process_equipment: 'Process Equipment',
  }[typeCode] || 'Equipment';
  const zone = subtitle.split(/\s+\+/)[0].trim();
  return `${typeLabel} ${tag} — ${zone}`;
}

function esc(s) {
  return String(s ?? '').replace(/'/g, "''");
}

function parseTsv(text) {
  const lines = text.split(/\r?\n/).filter((l) => l.trim());
  const rows = [];
  for (const line of lines) {
    const cols = line.split('\t');
    if (cols.length < 2) continue;
    const [tag, area_code, subtitle_location, drawing, level, project_site] = cols;
    if (!tag || tag === 'tag' || tag.trim() === '5') continue;
    if (!tag.trim()) continue;
    rows.push({
      tag: tag.trim(),
      area_code: (area_code || '').trim(),
      subtitle_location: (subtitle_location || '').trim(),
      drawing: (drawing || '').trim(),
      level: (level || '').trim(),
      project_site: (project_site || '').trim(),
    });
  }
  return rows;
}

function dedupeTags(rows) {
  const seen = new Map();
  const out = [];
  for (const row of rows) {
    let key = row.tag;
    if (seen.has(key)) {
      const prev = seen.get(key);
      const suffix = row.drawing || row.level || String(seen.size);
      key = `${row.tag}@${suffix}`;
      row.source_notes = `Duplicate tag; also at drawing ${prev.drawing} level ${prev.level}`;
    }
    row.unique_tag = key;
    seen.set(row.tag, row);
    out.push(row);
  }
  return out;
}

const raw = fs.readFileSync(inputPath, 'utf8');
const rows = dedupeTags(parseTsv(raw));
console.log(`Parsed ${rows.length} rows from ${inputPath}`);

const values = rows.map((r) => {
  const typeCode = inferTypeCode(r.tag);
  const { elevation_m, floor, zone: zoneFromLevel } = parseLevel(r.level, r.subtitle_location);
  const zone = zoneFromLevel || inferZone(r.subtitle_location);
  const building = inferBuilding(r.subtitle_location);
  const equipClass = inferEquipmentClass(r.tag);
  const props = {
    area_code: r.area_code,
    subtitle_location: r.subtitle_location,
    drawing: r.drawing,
    level_raw: r.level,
    elevation_m,
    zone,
    project_site: r.project_site,
    original_tag: r.tag,
  };
  if (equipClass) props.equipment_class = equipClass;
  if (typeCode === 'process_equipment') props.legend_pending = true;

  const locationCode = `BAH-${elevation_m ?? 0}-${zone.toUpperCase().replace(/_/g, '')}`;

  return `  ('${esc(r.unique_tag)}','${esc(r.tag)}','${esc(buildName(r.tag, r.subtitle_location, typeCode))}','${typeCode}','${building}','${floor}', '${esc(JSON.stringify({ ...props, location_code: locationCode }))}'::jsonb, 'batch-house-equipment-export', ${r.source_notes ? `'${esc(r.source_notes)}'` : 'NULL'}, 'medium', 'draft')`;
});

const header = `-- Migration 0012: Batch House process equipment → staging_objects
-- Source: batch-house-equipment.tsv (Ciner Glass Belgium / AGC legacy tags)
-- Legend: BA=Bin Activator, BE=Bucket Elevator, C=Conveyor, DCC=Pneumatic Hammer
-- Generated: ${new Date().toISOString().slice(0, 10)}

INSERT INTO public.staging_objects
  (tag, proposed_code, name, type_code, building_code, floor_name, properties, source_doc, source_notes, confidence, status)
VALUES
`;

const footer = `
ON CONFLICT (tag) DO UPDATE SET
  name = EXCLUDED.name,
  type_code = EXCLUDED.type_code,
  building_code = EXCLUDED.building_code,
  floor_name = EXCLUDED.floor_name,
  properties = EXCLUDED.properties,
  source_doc = EXCLUDED.source_doc,
  source_notes = COALESCE(EXCLUDED.source_notes, staging_objects.source_notes),
  updated_at = now();
`;

// Split into chunks of 100 for readability; single INSERT is fine for Postgres
const sql = header + values.join(',\n') + footer;
fs.writeFileSync(outSql, sql, 'utf8');
console.log(`Wrote ${outSql} (${rows.length} rows)`);
